frappe/modules/patch_handler.py

```diff
@@ -1,37 +1,58 @@
-# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
+# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and Contributors
 # License: MIT. See LICENSE
+""" Patch Handler.
+
+	This file manages execution of manaully written patches. Patches are script
+	that apply changes in database schema or data to accomodate for changes in the
+	code.
+
+	Ways to specify patches:
+		1. patches.txt file specifies patches that run before doctype schema
+		migration. This is only required when your patch needs access to data
+		that existed before schema migration.
+		2. patches_post.txt file is identical but runs patches after doctype schema is migrated.
+		3. small python commands can be passed added by starting ine with
+		`execute:` example: `execute:print("hello world")`
 """
-	Execute Patch Files
 
-	To run directly
+import time
 
-	python lib/wnf.py patch patch1, patch2 etc
-	python lib/wnf.py patch -f patch1, patch2 etc
+import frappe
 
-	where patch1, patch2 is module name
-"""
-import frappe, frappe.permissions, time
+class PatchError(Exception):
+	pass
+
+PATCH_FILES = {
+	"pre-migration": "patches.txt",
+	"post-migration": "patches_post.txt",
```

Comment:
The names are technically wrong (because the execution of patch is part of migration)... isn't it `pre-model-sync` & `post-model-sync`?

Reply:
ah yes. The flag names right?

Any suggestion on file name too? `patches.txt` can't be changed since it's used in many apps by now, `patches_post.txt` is okay?

Reply:
Would have preferred a single file for a patch... can we maintain a single JSON for a patch 😅 
something like

 **patch.json**
```JSON

{
	"pre-model-sync": [
		"frappe.patches.v13_0.rename_custom_client_script",
		"execute:frappe.reload_doc('core', 'doctype', 'role') #2017-05-23",
		"execute:frappe.reload_doc('core', 'doctype', 'user') #2017-10-27"
	],
	"post_model_sync": [
		"frappe.patches.v14_0.drop_data_import_legacy",
		"frappe.patches.v14_0.copy_mail_data #08.03.21"
	]
}
```
we can still support old `patch.txt` files?

Reply:
Single `patches.json` Sounds cool-er. 

1. We can keep `patch.txt` as "pre_model_sync" (current behaviour)
2. Using both `patches.txt` and `patches.json` shouldn't be allowed though. (no way for us to determine correct ordering)